README.md
Farsight DNSTable Manager
=========================

This tool is designed to maintain local copies of DNSTable filesets.
Given an URL it will periodically download a remote fileset descriptor and
all files referenced within.  It will maintain a local fileset file that
references all files present on the local system and correctly handles files
as they are merged over time (eg. from hourly to daily).

Installation
------------

Installation via Debian package requires the Farsight apt repository to be
installed.:

    apt-get install dnstable-manager
    # create /etc/dnstable-manager/dnstable-manager.yaml

    service dnstable-manager restart
    # OR
    apt-get install dnstable-manager-run # if you don't have systemd

Otherwise, you can use Python's package manager:

    python setup.py install
    # create /etc/dnstable-manager/dnstable-manager.yaml
    # configure your system to run dnstable-manager as a service

Running DNSTable Manager
------------------------

dnstable-manager works best when run as a systemd or runit service.  Both
configurations are included in the source tree and debian packages.  The
--config option is required.:

    usage: dnstable-manager [-h] [--config CONFIG] [--verbosity]
    
    Maintains local copies of remote filesets. Supports http, https, ftp, file,
    rsync, rsync+rsh uri schemas.
    
    optional arguments:
      -h, --help       show this help message and exit
      --config CONFIG  Path to configuration file.
      --verbosity, -v  Verbosity level. Repeat to increase.

Example:

    dnstable-manager --config /etc/dnstable-manager/dnstable-manager.yaml

Configuration
-------------

Configuration of dnstable-manager is done with a single YAML file specified
with the --config parameter on the command line.  It consists of three
sections:

```yaml
    manager:
        log_stream: 'stderr' or 'stdout'
        log_file: path to rotatable log file
        syslog: 'true' or 'false'
	syslog_facility: uppercase_name_of_facility
        log_level: one of 'CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG'
        clean_tempfiles: 'true' or 'false', cleans up stale temporary files at start
    downloader:
        max_downloads: integer, at least 3 recommended
        download_timeout: time in seconds
        retry_timeout: time in seconds
        tempdir: directory on filesystem with enough space, needed for rsync
        rsync_rsh: command line for RSYNC_RSH variable
	ssl_ca_file: ssl ca for validation
	ssl_keyfile: ssl client key
	ssl_certfile: ssl client cetificate
	ssl_ciphers: allows you to override the list of ssl ciphers to be used (default is considered secure at time of writing)
    filesets:
        name of fileset:
            uri: REQUIRED, remote uri to fileset, rsync+rsh protocol supported
            realm: optional HTTP authentication realm
            username: HTTP authentication username
	    password: HTTP authentication password
	    apikey: X-API-Key secret
            destination: REQUIRED, local destination dir of fileset
	    base: REQUIRED, prefix of fileset name (e.g. dns, dnssec, dns--com)
	    extension: REQUIRED, suffix of files in set (e.g. mtbl)
            frequency: REQUIRED, how often to download the fileset
            validator: validation command (filename is passed as argv[1])
            digest_required: require Digest header validation, set to false to disable
            minimal: optional boolean to enable base-full.fileset
```


debian/changelog
dnstable-manager (1.0.2) debian-farsightsec; urgency=medium

  * Refactor systemd service to own package

 -- Henry Stern <stern@fsi.io>  Tue, 06 Sep 2016 18:14:11 -0300

dnstable-manager (1.0.1) debian-farsightsec; urgency=medium

  * fileset: unlink was stripping extension from digest files

 -- Henry Stern <stern@fsi.io>  Wed, 13 Jul 2016 09:46:35 -0300

dnstable-manager (1.0.0) debian-farsightsec; urgency=medium

  * download,fileset: Add support for Digest header validation, caching
  * Add digest_required option to fileset to make Digest verification mandatory.
  * Add apikey support
  * dnstable-manager, fileset: add missing apikey glue
  * dnstable-manager: add missing digest_required config variable reference
  * dnstable-manager: temporarily make digest_required default to False

 -- Henry Stern <stern@fsi.io>  Mon, 11 Jul 2016 09:04:09 -0300

dnstable-manager (0.9.1) debian-farsightsec; urgency=medium

  * setup: change requires to install_requires
  * dnstable_manager: psutils compatibility between 1.2.0 and 4.2.0

 -- Henry Stern <stern@fsi.io>  Fri, 27 May 2016 16:55:36 -0300

dnstable-manager (0.9.0) debian-farsightsec; urgency=medium

  * README: add missing docs for download_timeout
  * Add LICENSE file
  * Add copyright header to all .py files
  * Add COPYRIGHT file for backported python ssl module
  * Add minimal/full distinction to filesets
  * Re-organize tests into own module.
  * fileset: add list_tempoary_files function
  * Add DnstableManager.clean_tempfiles function
  * dnstable-manager: add clean_tempfiles option, plumbing

 -- Henry Stern <stern@fsi.io>  Thu, 05 May 2016 15:16:11 -0300

dnstable-manager (0.8.0) debian-farsightsec; urgency=medium

  * Implement download timeouts

 -- Henry Stern <stern@fsi.io>  Tue, 19 Jan 2016 15:03:00 -0400

dnstable-manager (0.7) debian-farsightsec; urgency=medium

  * control: add pyflakes to build-dep
  * https: include backported ssl from 2.7.5 for older pythons
  * dnstable_manager: check if ssl cert and key files are readable
  * dnstable_manager: ensure that destination directories exist
  * dnstable-manager: default logging to stderr
  * download: add delete=True to NamedTemporaryFiles
  * dnstable-manager: replace joins with signal.pause
  * Make threads no longer daemonic
  * dnstable-manager: add termination log message
  * setup.py: depend on jsonschema >= 2.3.0

 -- Henry Stern <stern@fsi.io>  Fri, 06 Nov 2015 08:47:47 -0400

dnstable-manager (0.6) debian-farsightsec; urgency=medium

  * control: depend on jsonschema 2.3.0
  * dnstable-manager: make config a required argument

 -- Henry Stern <stern@fsi.io>  Thu, 05 Nov 2015 07:55:27 -0400

dnstable-manager (0.5) debian-farsightsec; urgency=medium

  * dnstable-manager: handle SIGTERM with sys.exit

 -- Henry Stern <stern@fsi.io>  Fri, 30 Oct 2015 14:15:06 -0300

dnstable-manager (0.4) debian-farsightsec; urgency=medium

  * fileset: add optional validator
  * Add MANIFEST.in

 -- Henry Stern <stern@fsi.io>  Mon, 05 Oct 2015 16:29:49 -0300

dnstable-manager (0.3) debian-farsightsec; urgency=medium

  * dnstable_manager: clean up exception handlers, handle httplib.HTTPException
  * https: convert ssl and socket exceptions to URLError
  * dnstable_manager: handle socket.error in load_remote_fileset
  * fileset: add FilesetException, check for matching content-length
  * download: Add Content-Length check
  * fileset: handle local files with unparseable names

 -- Henry Stern <stern@fsi.io>  Tue, 29 Sep 2015 09:53:32 -0300

dnstable-manager (0.2) debian-farsightsec; urgency=low

  * default_config: remove empty default log file
  * config: avoid duplicate destination/base
  * dnstable-manager: remove remnants of daemonization
  * fileset: allow unlinking to be disabled

 -- Henry Stern <stern@fsi.io>  Mon, 14 Sep 2015 17:08:39 +0000

dnstable-manager (0.1) debian-farsightsec; urgency=medium

  * Initial release.

 -- Henry Stern <stern@fsi.io>  Thu, 30 Jul 2015 13:56:19 -0300


tests/util.py
# Copyright (c) 2015 by Farsight Security, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import base64
from cStringIO import StringIO
import unittest

import dnstable_manager.util as du

def _iterfileobj(data, length):
    def t(self):
        src = StringIO(data)
        dst = StringIO()

        for chunk in du.iterfileobj(src, length=length):
            self.assertLessEqual(len(chunk), length)
            self.assertGreater(len(chunk), 0)
            dst.write(chunk)

        self.assertEqual(dst.getvalue(), data)
    return t

class TestIterFileObj(unittest.TestCase): pass

for length in (2, 32, 1024, 16*1024):
    data = base64.b64decode('qoiMsxenAEPPiamMzviaf13rCA6s2xtJrYjGBMZ=') * (length+1)
    assert(len(data) % length > 0)
    setattr(TestIterFileObj, 'test_iterfileobj(len={})'.format(length), _iterfileobj(data, length))


tests/rsync.py
# Copyright (c) 2015 by Farsight Security, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import print_function

import os
import subprocess
import urllib2
import unittest

from dnstable_manager.rsync import RsyncHandler

# TODO test attributes, validity of arguments?
class TestRsyncHandler(unittest.TestCase):
    file_data = 'test\ndata\n'
    fail_url = 'rsync://fail-url'

    def setUp(self):
        self.orig_check_call = subprocess.check_call
        subprocess.check_call = self.fake_check_call

    def tearDown(self):
        subprocess.check_call = self.orig_check_call

    def fake_check_call(self, argv, **kwargs):
        if argv[-2] == TestRsyncHandler.fail_url:
            raise urllib2.URLError('fail url')
        if not os.path.exists(argv[-1]):
            open(argv[-1], 'w').write(TestRsyncHandler.file_data)

    def test_urlopen(self):
        fp = urllib2.urlopen('rsync://localhost/test.txt')
        self.assertEqual(fp.read(), TestRsyncHandler.file_data)

    def test_urlopen_user(self):
        fp = urllib2.urlopen('rsync://foo@localhost/test.txt')
        self.assertEqual(fp.read(), TestRsyncHandler.file_data)

    def test_urlopen_fails(self):
        with self.assertRaises(urllib2.URLError):
            urllib2.urlopen(TestRsyncHandler.fail_url)

    def test_rsh(self):
        fp = urllib2.urlopen('rsync+rsh://foo@localhost/test.txt')
        self.assertEqual(fp.read(), TestRsyncHandler.file_data)

    def test_cookiemonster(self):
        handler = RsyncHandler()
        attrs = ['a=b']
        handler.do_rsync('rsync://localhost/test.txt', attrs=attrs)
        self.assertItemsEqual(attrs, ['a=b'])


tests/digest.py
# Copyright (c) 2015 by Farsight Security, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.import base64

import base64
from cStringIO import StringIO
import hashlib
import textwrap
import unittest

import dnstable_manager.digest as dd

def _test_digest(algorithm, passed_algorithm, invalid=False):
    def t(self):
        text = textwrap.dedent('''\
                fwKMyjBG1gjQqEUpcL1UFj3e9tyjVuD8vE3WtU6CjpyuV
                nn78zKqeuYc0H8dLopt11DHRzlBBBpowwVZMYjl3BW8wm
                Kj0ovutecEjWrQsjygPwGl7uKLMAxhyzYp9udfJtxAPoD
                T3Uk3D4yaexKDeKkuSYaIIGk0safQMrBoP6UYMT3B9rJM
                DoDe0vQxqfoOSmPgfGc4bBgau6mF9yc8Bg804zjnbKXbW
                uXytE27bdUf942xsf8tTVW9xwhfsIMexjoE5hazQWmDCd
                RHYOiYmaDTlmI4mwknzLEPPVJVXwOUPU8Sy1BXV3ZQ3Ur
                YNb2o5vWbWy8QoCjpzIeFn4TakzJm71bOy76GiI9eOAcB
                TJunbclmdi9hRk4TO5KPdLS3XucJcS0xiigPbOj4OLLMb
                zIO75BYeS0MyjQPNplYvsNLAuLHUAZWBVQRxIdU4xoaFI
                xALV4Ryb94Z4C1gMy4r8MK6dp8N5dYRKfSRRAQxx2ZULo
                aEA8Wyx2uMexsJVbi4mqpOt4hVh3l1ZcM18kZByFEJjKt
                fVvIVE3YYujPaLhXE7JbbaAPoJsrMvB4BBrIABjmAFNSf
                mobzY6nd0RGxJtJYUd6kZaNsxFm7TGTQOB5YTXZ6223ev
                tjXBOwsL3HDCRYEkaPwYlRkPoR2MmnF4ONTXjjK6YPrLW
                YAZtHXDNrLBlmjQWGAP1GQ669bE77HQ58tlXuzMSvqiQt
                qmN0s9oyaUVokcTBFIJPYorLLMhhOwaF6h72QRZWDBIgG
                R5hqwvVPhBUcmmVheKTOv18RzS91TqaOL94VZT8N0lv0M
                4O5KZpkqiy1YpOS61BvCFr7IWdwmI98QKdbdpMLUAydXQ
                OFVG60va76U2PyphDe22sBRbmwnzwjGf8yNXzb2nhsMhb
                GB6EKRy7JadeKpzqCg7iJdqUqOoQbgvgzMJDYRuTiQJtu
                hB9uyyPbWGb4sc1pzYTjED0kLSl4PjImdvuasWWfegwYr
                knVt0HLP1PypZxwxcpsi8rdlLeupMHK2eGNLgvOIaomWB
                ARVRKS3KHVuGCD36CfSNzWIorOfjF3j53Bfb9Qk4XJWO7\
                        ''')

        if invalid:
            digest = 'INVALID'
        else:
            digest = base64.b64encode(hashlib.new(algorithm, text).digest())

        try:
            for line_no,(expected,actual) in enumerate(zip(StringIO(text), dd.check_digest(StringIO(text), passed_algorithm, digest))):
                self.assertEqual(expected, actual, msg='Line {}: {!r} != {!r})'.format(line_no+1, expected, actual))
        except dd.DigestError:
            if invalid:
                pass
            else:
                raise
    return t

def _test_digest_extension(expected, algorithm, invalid=False):
    def t(self):
        extension = dd.digest_extension(algorithm)
        try:
            self.assertEqual(extension, expected)
        except dd.DigestError:
            if invalid:
                pass
            else:
                raise
    return t

class TestDigest(unittest.TestCase):
    test_no_algorithm = _test_digest('sha256', None)
    test_invalid_algorithm = _test_digest('sha256', 'MD5')

for bits in (224, 256, 384, 512):
    setattr(TestDigest, 'test_digest(sha{})'.format(bits), _test_digest('sha{}'.format(bits), 'sha{}'.format(bits)))
    setattr(TestDigest, 'test_digest(SHA-{})'.format(bits), _test_digest('sha{}'.format(bits), 'SHA-{}'.format(bits)))
    setattr(TestDigest, 'test_digest_invalid(sha{})'.format(bits), _test_digest('sha{}'.format(bits), 'sha{}'.format(bits), invalid=True))
    setattr(TestDigest, 'test_digest_extension(sha{})'.format(bits), _test_digest_extension('sha{}'.format(bits), 'sha{}'.format(bits)))
    setattr(TestDigest, 'test_digest_extension(SHA-{})'.format(bits), _test_digest_extension('sha{}'.format(bits), 'SHA-{}'.format(bits)))


dnstable_manager/https.py
# Copyright (c) 2015 by Farsight Security, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import urllib2
import httplib
import logging
import socket

logger = logging.getLogger(__name__)

try:
    import ssl
    ssl.CertificateError
except AttributeError:
    logger.debug('Using backported ssl.py')
    import backports.ssl as ssl

ca_file = '/etc/ssl/certs/ca-certificates.crt'
keyfile = None
certfile = None
ciphers = 'EECDH+ECDSA+AESGCM:EECDH+aRSA+AESGCM:EECDH+ECDSA+SHA384:EECDH+ECDSA+SHA256:EECDH+aRSA+SHA384:EECDH+aRSA+SHA256:!EECDH+aRSA+RC4:EECDH:EDH+aRSA:!RC4:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!SRP:!DSS:@STRENGTH'

class HTTPSConnection(httplib.HTTPConnection):
    default_port = httplib.HTTPS_PORT

    def __init__(self, *args, **kwargs):
        httplib.HTTPConnection.__init__(self, *args, **kwargs)

    def connect(self):
        try:
            sock = socket.create_connection((self.host, self.port), self.timeout, self.source_address)
        except socket.error as e:
            raise urllib2.URLError(e)
        if self._tunnel_host:
            self.sock = sock
            self._tunnel()
        try:
            self.sock = ssl.wrap_socket(sock, ssl_version=ssl.PROTOCOL_TLSv1,
                    ca_certs=ca_file, cert_reqs=ssl.CERT_REQUIRED,
                    keyfile=keyfile, certfile=certfile, ciphers=ciphers)
        except (ssl.SSLError, ssl.CertificateError, socket.error) as e:
            raise urllib2.URLError(e)

class HTTPSHandler(urllib2.HTTPSHandler):
    def https_open(self, req):
        return self.do_open(HTTPSConnection, req)
    handler_order = urllib2.HTTPSHandler.handler_order - 1


dnstable_manager/digest.py
# Copyright (c) 2015 by Farsight Security, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import base64
import hashlib
import logging

logger = logging.getLogger(__name__)

class DigestError(Exception): pass

def check_digest(iterator, algorithm, digest):
    logger.debug('algorithm={}, checksum={}'.format(algorithm, digest))

    if algorithm is None:
        logger.debug('No algorithm provided. Skipping digest check.')
        for chunk in iterator:
            yield chunk
        return
    elif algorithm.lower() in ('sha-224', 'sha224'):
        digest_obj = hashlib.sha224()
    elif algorithm.lower() in ('sha-256', 'sha256'):
        digest_obj = hashlib.sha256()
    elif algorithm.lower() in ('sha-384', 'sha384'):
        digest_obj = hashlib.sha384()
    elif algorithm.lower() in ('sha-512', 'sha512'):
        digest_obj = hashlib.sha512()
    else:
        logger.debug('Unsupported algorithm: {}'.format(algorithm))
        for chunk in iterator:
            yield chunk
        return

    for chunk in iterator:
        digest_obj.update(chunk)
        yield chunk

    real_digest = base64.b64encode(digest_obj.digest())
    if real_digest != digest:
        raise DigestError('Digest mismatch: {} != {}'.format(real_digest, digest))

def digest_extension(algorithm):
    if algorithm.lower() in ('sha-224', 'sha224'):
        return 'sha224'
    elif algorithm.lower() in ('sha-256', 'sha256'):
        return 'sha256'
    elif algorithm.lower() in ('sha-384', 'sha384'):
        return 'sha384'
    elif algorithm.lower() in ('sha-512', 'sha512'):
        return 'sha512'
    else:
        raise DigestError('Unknown algorithm: {}'.format(algorithm))

DIGEST_EXTENSIONS = ('sha224', 'sha256', 'sha384', 'sha512')


dnstable_manager/fileset.py
# Copyright (c) 2015 by Farsight Security, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import print_function
import calendar
import datetime
import errno
import glob
import logging
import os
import subprocess
import tempfile
import time
import urllib
import urllib2

from .digest import DigestError, check_digest, DIGEST_EXTENSIONS

logger = logging.getLogger(__name__)
disable_unlink = False

class FilesetError(Exception): pass

class ParseError(FilesetError): pass

class ValidationFailed(FilesetError): pass

def parse_datetime(s):
    """
    Parse a Y/M/D/H string and return a datetime object. The length of the
    input string implicitly identifies the time frequency (Y/M/D/H).

    Examples:

    >>> parse_datetime('20060102.1500')     # hour
    datetime.datetime(2006, 1, 2, 15, 0)
    >>> parse_datetime('20060102')          # day
    datetime.datetime(2006, 1, 2, 0, 0)
    >>> parse_datetime('200601')            # month
    datetime.datetime(2006, 1, 1, 0, 0)
    >>> parse_datetime('2006')              # year
    datetime.datetime(2006, 1, 1, 0, 0)
    >>>
    """
    fmt_times = ((13, '%Y%m%d.%H%M'), (8, '%Y%m%d'), (6, '%Y%m'), (4, '%Y'))
    for len_fmt, fmt in fmt_times:
        try:
            if len(s) == len_fmt:
                return datetime.datetime.utcfromtimestamp(calendar.timegm(time.strptime(s, fmt)))
        except ValueError:
            pass
    raise ParseError("Time data '{}' does not match any of the time formats".format(s))

def relative_uri(uri, fn):
    path,query = urllib.splitquery(uri)
    path,attrs = urllib.splitattr(path)
    if fn.startswith('/'):
        scheme,_,path = path.partition(':')
        host,path = urllib.splithost(path)
        new_uri = '{}://{}/{}'.format(scheme, host, fn[1:])
    else:
        parent = path.rpartition('/')[0]
        new_uri = '{}/{}'.format(parent, fn)
    if attrs:
        new_uri = '{};{}'.format(new_uri, ';'.join(attrs))
    return new_uri

def compute_overlap(files):
    years = set()
    quarters = set()
    months = set()
    weeks = set()
    days = set()
    hours = set()
    decaminutes = set()

    # Helper functions year_overlap(), month_overlap(), and day_overlap()
    # that check if their timestamp parameter is overlapped by a year,
    # month, or day timestamp.
    def year_overlap(dt):
        for y in years:
            if dt.year == y.year:
                return True
        return False

    def quarter_overlap(dt):
        if year_overlap(dt):
            return True
        dt_q = int((dt.month-1) / 3) * 3 + 1
        for q in quarters:
            if dt.year == q.year and dt_q == q.month:
                return True
        return False

    def month_overlap(dt):
        if quarter_overlap(dt):
            return True
        for m in months:
            if dt.year == m.year and dt.month == m.month:
                return True
        return False

    def week_overlap(dt):
        if month_overlap(dt):
            return True
        dt_wd = int((dt.day-1) / 7) * 7 + 1
        for w in weeks:
            if dt.year == w.year and dt.month == w.month and dt_wd == w.day:
                return True
        return False

    def day_overlap(dt):
        if week_overlap(dt):
            return True
        for d in days:
            if dt.year == d.year and dt.month == d.month and dt.day == d.day:
                return True
        return False

    def hour_overlap(dt):
        if day_overlap(dt):
            return True
        for h in hours:
            if dt.year == h.year and dt.month == h.month and dt.day == h.day and dt.hour == h.hour:
                return True
        return False

    def decaminute_overlap(dt):
        if hour_overlap(dt):
            return True
        dt_dm = int(dt.minute / 10) * 10
        for d in decaminutes:
            if dt.year == d.year and dt.month == d.month and dt.day == d.day and dt.hour == d.hour and dt_dm == d.minute:
                return True
        return False

    for f in sorted(files):
        if f.tl == 'Y':
            years.add(f.datetime)
        elif f.tl == 'M':
            if year_overlap(f.datetime):
                yield f
            else:
                months.add(f.datetime)
        elif f.tl == 'M':
            if quarter_overlap(f.datetime):
                yield f
            else:
                months.add(f.datetime)
        elif f.tl == 'W':
            if month_overlap(f.datetime):
                yield f
            else:
                weeks.add(f.datetime)
        elif f.tl == 'D':
            if week_overlap(f.datetime):
                yield f
            else:
                days.add(f.datetime)
        elif f.tl == 'H':
            if day_overlap(f.datetime):
                yield f
            else:
                hours.add(f.datetime)
        elif f.tl == 'X':
            if hour_overlap(f.datetime):
                yield f
            else:
                decaminutes.add(f.datetime)
        elif f.tl == 'm':
            if decaminute_overlap(f.datetime):
                yield f

class File(object):
    """
    Helper class for Fileset which wraps the parsing of Y/M/D/H filenames.
    """

    _valid_tl = ('Y', 'Q', 'M', 'W', 'D', 'H', 'X', 'm')

    def __init__(self, name, dname=None, uri=None, apikey=None, validator=None, digest_required=True):
        self.name = name
        self.dname = dname
        self.uri = uri
        self.apikey = apikey
        self.validator = validator
        self._init_tl()
        self._init_datetime()
        self.digest_required = digest_required

    def __repr__(self):
        return '<File %r, tl %r, %r, dir %r, uri %r>' % (self.name, self.tl, self.datetime, self.dname, self.uri)

    def _init_tl(self):
        try:
            tl = self.name.split('.')[-2]
        except IndexError:
            raise ParseError('Unable to parse time letter from file name {}'.format(self.name))

        if not tl in File._valid_tl:
            raise ParseError('Time letter {} not in valid set {}'.format(tl, File._valid_tl))
        self.tl = tl

    def _init_datetime(self):
        base_name = os.path.basename(self.name)

        # Strip off the leading filename component and the two trailing filename components.
        # E.g., 'dns.20060102.1500.H.mtbl' -> '20060102.1500'
        # E.g., 'dns.2006.Y.mtbl' -> '2006'
        datetime_string = '.'.join(base_name.split('.')[1:-2])

        try:
            self.datetime = parse_datetime(datetime_string)
        except ParseError as e:
            raise ParseError('Unable to extract datetime from filename {}: {}'.format(base_name, e))

    def __cmp__(self, other):
        return cmp(File._valid_tl.index(self.tl), File._valid_tl.index(other.tl)) or cmp(self.datetime, other.datetime) or cmp(self.name, other.name)

    def __hash__(self):
        return hash(self.tl) ^ hash(self.datetime) ^ hash(self.name)

    def target(self):
        if self.dname:
            return os.path.join(self.dname, self.name)
        else:
            return self.name

    def validate(self, filename=None):
        if filename is None:
            filename = self.target()

        if self.validator:
            stdout = tempfile.TemporaryFile()
            stderr = tempfile.TemporaryFile()
            try:
                logger.info('Validating {}'.format(filename))
                subprocess.check_call([self.validator, filename], stdout=stdout, stderr=stderr)
                stdout.seek(0)
                logger.debug('stdout: {}'.format(stdout.read()))
                stderr.seek(0)
                logger.debug('stderr: {}'.format(stderr.read()))
            except subprocess.CalledProcessError:
                stderr.seek(0)
                raise ValidationFailed('Validation of {} failed: {}'.format(filename, stderr.read()))

class Fileset(object):
    def __init__(self, uri, dname, base='dns', extension='mtbl', apikey=None, validator=None, digest_required=True, timeout=None):
        """
        Create a new Fileset object.

        'dname' is the destination directory containing files.
        'base' is the filename prefix (e.g., "dns", "dnssec").
        'extension' is the filename suffix (e.g., "mtbl").

        The Fileset will be initialized with all files named like
        '{dname}/{base}.*.[YMWDHXm].{extension}'.
        """

        if not os.path.isdir(dname):
            raise Exception('Unable to open fileset directory: {}'.format(dname))

        self.uri = uri
        self.dname = dname
        self.base = base
        self.extension = extension
        self.apikey = apikey
        self.validator = validator
        self.digest_required = digest_required
        self.timeout = timeout

        self.all_local_files = None
        self.minimal_local_files = None
        self.load_local_fileset()
        self.remote_files = set(self.all_local_files)
        self.pending_deletions = set()

    def load_local_fileset(self):
        g_expr = '{}/{}.*.[YQMWDHXm].{}'.format(self.dname, self.base, self.extension)
        new_local_files = set()
        for fname in glob.glob(g_expr):
            try:
                new_local_files.add(File(os.path.basename(fname), validator=self.validator, apikey=self.apikey, digest_required=self.digest_required))
            except ParseError as e:
                logger.debug('Error parsing filename \'{}\': {}'.format(fname, str(e)))
        self.all_local_files = set(new_local_files)
        self.minimal_local_files = set(new_local_files)

    def prune_obsolete_files(self, minimal=True):
        obsolete_files = self.minimal_local_files.difference(self.remote_files).difference(compute_overlap(self.minimal_local_files.union(self.remote_files)))

        if not minimal:
            obsolete_files.update(set(compute_overlap(self.minimal_local_files)).difference(compute_overlap(self.remote_files)))

        self.all_local_files.difference_update(obsolete_files)
        self.minimal_local_files.difference_update(obsolete_files)
        self.pending_deletions.update(obsolete_files)

    def prune_redundant_files(self, minimal=True):
        redundant_files = set(compute_overlap(self.minimal_local_files))
        self.minimal_local_files.difference_update(redundant_files)
        if minimal:
            self.all_local_files.difference_update(redundant_files)
            self.pending_deletions.update(redundant_files)

    def get_fileset_name(self, minimal=True):
        if not minimal:
            return os.path.join(self.dname, self.base + '-full.fileset')
        return os.path.join(self.dname, self.base + '.fileset')

    def _write_fileset(self, fileset, fileset_fname):
        # Read the old fileset, if it exists.
        try:
            old_fileset = set(line.rstrip() for line in open(fileset_fname))
        except IOError as e:
            if e.errno == errno.ENOENT:
                logger.debug('Fileset {} does not exist.  Starting with a blank one.'.format(fileset_fname))
                old_fileset = set()
            else:
                raise

        if old_fileset.symmetric_difference(f.name for f in fileset) or not os.path.exists(fileset_fname):
            logger.debug('Fileset {} has changed'.format(fileset_fname))
            with tempfile.NamedTemporaryFile(prefix='.{}.'.format(os.path.basename(fileset_fname)), dir=os.path.dirname(fileset_fname), delete=True) as out:
                for f in sorted(fileset):
                    print (f.name, file=out)
                out.file.close()
                os.chmod(out.name, 0o644)
                os.rename(out.name, fileset_fname)
                out.delete = False

    def write_local_fileset(self, minimal=True):
        if not minimal:
            fileset = self.all_local_files
        else:
            fileset = self.minimal_local_files

        self._write_fileset(fileset, self.get_fileset_name(minimal=minimal))

    def purge_deleted_files(self):
        for f in sorted(self.pending_deletions):
            fn = os.path.join(self.dname, f.name)
            logger.info('Unlinking {}'.format(fn))
            try:
                if not disable_unlink:
                    os.unlink(fn)
                    for extension in DIGEST_EXTENSIONS:
                        digest_fn = '{}.{}'.format(fn, extension)
                        try:
                            os.unlink(digest_fn)
                            logger.info('Unlinked digest file {}'.format(digest_fn))
                        except OSError as e:
                            if e.errno != errno.ENOENT:
                                logger.error('Could not unlink digest file {}: {}'.format(digest_fn, e))
            except OSError as e:
                if e.errno == errno.ENOENT:
                    logger.error('File vanished {}'.format(fn))
                else:
                    raise
            self.pending_deletions.remove(f)

    def load_remote_fileset(self):
        logger.info('Retrieving {}'.format(self.uri))
        req = urllib2.Request(self.uri)
        if self.apikey:
            req.add_header('X-API-Key', self.apikey)
        fp = urllib2.urlopen(req, timeout=self.timeout)
        new_remote_files = set()
        read_len = 0
        algorithm = None
        digest = None

        if 'Digest' in fp.headers:
            algorithm,_,digest = fp.headers['Digest'].partition('=')

        try:
            for fname in check_digest(fp, algorithm, digest):
                read_len += len(fname)
                fname = fname.rstrip()

                if os.path.basename(fname) != fname:
                    logger.warning('Skipping {}.  Not a basename.'.format(fname))
                    continue
                if not fname.startswith('{}.'.format(self.base)):
                    logger.warning('Skipping {}.  Base is not {}.'.format(fname, self.base))
                    continue
                if not fname.endswith('.{}'.format(self.extension)):
                    logger.warning('Skipping {}.  Extensions is not {}.'.format(fname, self.extension))
                    continue

                new_remote_files.add(File(fname, dname=self.dname, uri=relative_uri(self.uri, fname), validator=self.validator, apikey=self.apikey, digest_required=self.digest_required))
        except DigestError as e:
            raise FilesetError(e)

        if 'Content-Length' in fp.headers:
            try:
                if read_len != int(fp.headers['Content-Length']):
                    raise FilesetError('Fileset content length mismatch: {} != {}'.format(read_len, fp.headers['Content-Length']))
            except ValueError:
                logger.debug('Skipping Content-Length check, unparsable header')
        else:
            logger.debug('Skipping Content-Length check')

        self.remote_files = new_remote_files

    def missing_files(self):
        return self.remote_files.difference(self.all_local_files)

    def list_temporary_files(self):
        return glob.glob(os.path.join(self.dname, '.{}.*.{}.*'.format(self.base, self.extension)))
